pkg/controller/clusterpool/collections.go

```diff
@@ -371,20 +373,102 @@ func (cds *cdCollection) SyncClaimAssignments(c client.Client, claims *claimColl
 	cds.MakeNotAssignable(invalidCDs...)
 }
 
+func removeCDsFromSlice(slice *[]*hivev1.ClusterDeployment, cdNames ...string) {
+	for _, cdName := range cdNames {
+		for i, cd := range *slice {
+			if cd.Name == cdName {
+				copy((*slice)[i:], (*slice)[i+1:])
+				*slice = (*slice)[:len(*slice)-1]
+			}
+		}
+	}
+}
+
 // MakeNotAssignable idempotently removes the named ClusterDeployments from the assignable list of the
 // cdCollection, so they are no longer considered for assignment. They still count against pool
 // capacity. Do this to a broken ClusterDeployment -- e.g. one that
 // - is assigned to the wrong claim, or a claim that doesn't exist
 // - can't be synced with its claim for whatever reason in this iteration (e.g. Update() failure)
 func (cds *cdCollection) MakeNotAssignable(cdNames ...string) {
-	for _, cdName := range cdNames {
-		for i, cd := range cds.assignable {
-			if cd.Name == cdName {
-				copy((cds.assignable)[i:], (cds.assignable)[i+1:])
-				cds.assignable = cds.assignable[:len(cds.assignable)-1]
-			}
+	removeCDsFromSlice(&cds.assignable, cdNames...)
+}
+
+// Delete deletes the named ClusterDeployment from the server, moving it from Assignable() to
+// Deleting()
+func (cds *cdCollection) Delete(c client.Client, cdName string) error {
+	cd := cds.ByName(cdName)
+	if cd == nil {
+		return errors.New(fmt.Sprintf("No such ClusterDeployment %s to delete. This is a bug!", cdName))
+	}
+	if err := c.Delete(context.Background(), cd); err != nil {
+		return err
+	}
+	cds.deleting = append(cds.deleting, cd)
+	// Remove from any of the other lists it might be in
+	removeCDsFromSlice(&cds.assignable, cdName)
+	removeCDsFromSlice(&cds.installing, cdName)
+	removeCDsFromSlice(&cds.assignable, cdName)
+	removeCDsFromSlice(&cds.markedForDeletion, cdName)
+	return nil
+}
+
+// setCDsCurrentCondition idempotently sets the ClusterDeploymentsCurrent condition on the
+// ClusterPool according to whether all unassigned CDs have the same PoolVersion as the pool.
+func setCDsCurrentCondition(c client.Client, cds *cdCollection, clp *hivev1.ClusterPool, poolVersion string) error {
+	// CDs with mismatched poolVersion
+	mismatchedCDs := make([]string, 0)
+	// CDs with no poolVersion
+	unknownCDs := make([]string, 0)
+
+	for _, cd := range append(cds.assignable, cds.installing...) {
+		annotations := cd.Annotations
+		if annotations == nil {
+			annotations = map[string]string{}
+		}
+		if cdPoolVersion, ok := annotations[constants.ClusterDeploymentPoolSpecHashAnnotation]; !ok {
+			// Annotation is either missing or empty. This could be due to upgrade (this CD was
```

Comment:
we are not checking for the value to be empty in the conditional? how can we be sure that annotation is empty?